components/identity-core/org.wso2.carbon.identity.core/src/main/java/org/wso2/carbon/identity/core/model/ServiceURL.java

```diff
@@ -0,0 +1,174 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.core.model;
+
+import org.apache.commons.collections.MapUtils;
+import org.apache.commons.lang.StringUtils;
+import org.wso2.carbon.identity.core.util.IdentityCoreConstants;
+
+import java.util.Map;
+
+/**
+ * Service URL representation.
+ */
+public class ServiceURL {
+
+    private String protocol;
+    private String hostName;
+    private Integer port;
+    private String proxyContext;
+    private String webContext;
+    private String urlContext;
+    private Map<String, String> parameters;
+    private Map<String, String> fragments;
+
+    public ServiceURL(String protocol, String hostName, Integer port, String proxyContext, String webContext,
+                      String urlContext, Map<String, String> parameters, Map<String, String> fragments) {
+
+        this.protocol = protocol;
+        this.hostName = hostName;
+        this.port = port;
+        this.proxyContext = proxyContext;
+        this.webContext = webContext;
+        this.urlContext = urlContext;
+        this.parameters = parameters;
+        this.fragments = fragments;
+    }
+
+    public String getProtocol() {
+
+        return protocol;
+    }
+
+    public String getHostName() {
+
+        return hostName;
+    }
+
+    public Integer getPort() {
+
+        return port;
+    }
+
+    public String getProxyContext() {
+
+        return proxyContext;
+    }
+
+    public String getWebContext() {
+
+        return webContext;
+    }
+
+    public String getUrlContext() {
+
+        return urlContext;
+    }
+
+    public Map<String, String> getParameters() {
+
+        return parameters;
+    }
+
+    public Map<String, String> getFragments() {
+
+        return fragments;
+    }
+
+    private void appendContextToUri(StringBuilder serverUrl, String contextPath) {
+
+        if (StringUtils.isNotBlank(contextPath)) {
+            if (contextPath.trim().charAt(0) != '/') {
+                serverUrl.append("/").append(contextPath.trim());
+            } else {
+                serverUrl.append(contextPath.trim());
+            }
+        }
+    }
+
+    private void appendParamsToUri(StringBuilder serverUrl, Map<String, String> parameters) {
+
+        if (MapUtils.isNotEmpty(parameters)) {
+            for (Map.Entry<String, String> entry : parameters.entrySet()) {
+                serverUrl.append("?").append(entry.getKey()).append("=").append(entry.getValue());
+            }
+        }
+    }
+
+    /**
+     * Concatenate the protocol, host name, port, proxy context path, web context root and url context to return the
+     * absolute URL.
+     *
+     * @return The absolute URL from the {@link ServiceURL} instance.
+     */
+    public String getAbsoluteUrl() {
```

Comment:
What if we build this at once with the constructor. Or build at first invocation.
There's no way these fields can change later as this class is immutable